src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java

```diff
@@ -347,56 +301,66 @@ public boolean isPaused() {
+        boolean changed = false;
         boolean ending = false;
-        do {
-            changed = false;
-            for (CpsThread t : threads.values().toArray(new CpsThread[threads.size()])) {
-                if (t.isRunnable()) {
-                    Outcome o = t.runNextChunk();
-                    if (o.isFailure()) {
-                        assert !t.isAlive();    // failed thread is non-resumable
-
-                        // workflow produced an exception
-                        Result result = Result.FAILURE;
-                        Throwable error = o.getAbnormal();
-                        if (error instanceof FlowInterruptedException) {
-                            result = ((FlowInterruptedException) error).getResult();
-                        }
-                        execution.setResult(result);
-                        t.head.get().addAction(new ErrorAction(error));
+        boolean stillRunnable = false;
+
+        // TODO: maybe instead of running all the thread, run just one thread in round robin
+        for (CpsThread t : threads.values().toArray(new CpsThread[threads.size()])) {
+            if (t.isRunnable()) {
+                Outcome o = t.runNextChunk();
+                if (o.isFailure()) {
+                    assert !t.isAlive();    // failed thread is non-resumable
+
+                    // workflow produced an exception
+                    Result result = Result.FAILURE;
+                    Throwable error = o.getAbnormal();
+                    if (error instanceof FlowInterruptedException) {
+                        result = ((FlowInterruptedException) error).getResult();
                     }
+                    execution.setResult(result);
+                    t.head.get().addAction(new ErrorAction(error));
+                }
 
-                    if (!t.isAlive()) {
-                        LOGGER.fine("completed " + t);
-                        t.fireCompletionHandlers(o); // do this after ErrorAction is set above
+                if (!t.isAlive()) {
+                    LOGGER.fine("completed " + t);
+                    t.fireCompletionHandlers(o); // do this after ErrorAction is set above
 
-                        threads.remove(t.id);
-                        if (threads.isEmpty()) {
-                            execution.onProgramEnd(o);
-                            ending = true;
-                        }
+                    threads.remove(t.id);
+                    if (threads.isEmpty()) {
+                        execution.onProgramEnd(o);
+                        ending = true;
                     }
-
-                    changed = true;
+                } else {
+                    stillRunnable |= t.isRunnable();
                 }
-            }
 
-            doneSomeWork |= changed;
-        } while (changed);
+                changed = true;
+            }
+        }
 
-        if (doneSomeWork) {
+        if (changed) {
             saveProgram();
```

Comment:
Caused some regressions in programs which were legitimately broken. I am more concerned that the constant saving may have caused some performance regressions. Perhaps we need to throttle saves.